warehouse/oidc/services.py

```diff
@@ -287,11 +310,30 @@ def find_publisher(
             publisher = find_publisher_by_issuer(
                 self.db, self.issuer_url, signed_claims, pending=pending
             )
+
+            jwt_token_identifier: str | None = signed_claims.get("jti")
+            # jti is in the __preverified_claims__ set, so if it was present,
+            # it was already checked
+            if pending is False and jwt_token_identifier:
+                if self.token_identifier_exists(jwt_token_identifier):
+                    self.metrics.increment(
+                        "warehouse.oidc.reused_token",
+                        tags=metrics_tags,
+                    )
+                    raise ReusedTokenError("JWT Token already used to mint a token.")
+
             publisher.verify_claims(signed_claims)
             self.metrics.increment(
                 "warehouse.oidc.find_publisher.ok",
                 tags=metrics_tags,
             )
+
+            if pending is False and jwt_token_identifier:
+                # Of note, exp is coming from a trusted source here,
+                # so we don't validate it
+                expiration = typing.cast(int, signed_claims.get("exp"))
+                self.store_jwt_identifier(jwt_token_identifier, expiration)
```

Comment:
I could be convinced otherwise, but I would expect us to actually store a JWT identifier as "used" only after it's been used -- specifically, when we've finished minting a token. As is, this invalidates tokens every time we see them, not necessarily every time we use them.

Reply:
Hmm, good point -- yeah, I think we should perform `store_jwt_identifier` at the point of minting, rather than here.

Reply:
👍